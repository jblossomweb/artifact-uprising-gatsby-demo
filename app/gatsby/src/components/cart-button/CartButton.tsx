import React from "react"
import Button from "@material-ui/core/Button"
import Input from "@material-ui/core/Input"
import AddIcon from "@material-ui/icons/Add"
import RemoveIcon from "@material-ui/icons/Remove"
import ShoppingCartIcon from "@material-ui/icons/ShoppingCart"
import AddShoppingCartIcon from "@material-ui/icons/AddShoppingCart"

import * as Style from "./CartButton.style"

export interface Props {
  qtyInCart: number
  addToCart: () => void
  subtractFromCart: () => void
  setQty: (qty: number) => void
}

const CartButton: React.FC<Props> = ({
  qtyInCart,
  addToCart,
  subtractFromCart,
  setQty,
}) => {
  if (qtyInCart < 0) qtyInCart = 0
  return qtyInCart ? (
    <Style.Input>
      <Input
        data-testid="qtyInput"
        type="number"
        disableUnderline
        value={qtyInCart}
        inputProps={{ min: 0, max: 99 }}
        startAdornment={
          <Button
            data-testid="subtractButton"
            variant="outlined"
            onClick={() => {
              subtractFromCart()
            }}
          >
            <RemoveIcon fontSize="small" />
          </Button>
        }
        endAdornment={
          <Button
            data-testid="addButton"
            variant="outlined"
            onClick={() => {
              addToCart()
            }}
          >
            <AddIcon fontSize="small" />
          </Button>
        }
        onChange={({ target: { value } }) => {
          setQty(Number(value))
        }}
      />
      &nbsp;
      <ShoppingCartIcon fontSize="small" />
    </Style.Input>
  ) : (
    <Button
      data-testid="addToCartButton"
      variant="outlined"
      onClick={() => {
        addToCart()
      }}
    >
      Add to Cart&nbsp;
      <AddShoppingCartIcon />
    </Button>
  )
}

export default CartButton
