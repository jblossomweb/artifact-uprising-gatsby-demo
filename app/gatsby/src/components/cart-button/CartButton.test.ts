import storyMounter from "../../../utils/story-mounter"
import snapshotTests from "../../../utils/story-snapshot-tests"
import { makeStories } from "./CartButton.stories"
import CartButton from "./CartButton"

const mockActions = {
  addToCart: jest.fn(),
  subtractFromCart: jest.fn(),
  setQty: jest.fn(),
}

const stories = makeStories(mockActions)

const mountedStories = storyMounter(stories, CartButton, component => ({
  addToCartButton: component.find('button[data-testid="addToCartButton"]'),
  qtyInput: component.find('[data-testid="qtyInput"]').find("input"),
  addButton: component.find('button[data-testid="addButton"]'),
  subtractButton: component.find('button[data-testid="subtractButton"]'),
}))

describe("components/cart-button", () => {
  snapshotTests(stories)

  describe("qty in cart, click add button", () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it("calls addToCart", () => {
      const story = mountedStories["qty in cart"]
      const { addButton } = story.elements
      const { addToCart } = mockActions
      expect(addButton.length).toBe(1)
      expect(addToCart).not.toHaveBeenCalled()
      addButton.simulate("click")
      expect(addToCart).toHaveBeenCalled()
    })
  })

  describe("qty in cart, click subtract button", () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it("calls subtractFromCart", () => {
      const story = mountedStories["qty in cart"]
      const { subtractButton } = story.elements
      const { subtractFromCart } = mockActions
      expect(subtractButton.length).toBe(1)
      expect(subtractFromCart).not.toHaveBeenCalled()
      subtractButton.simulate("click")
      expect(subtractFromCart).toHaveBeenCalled()
    })
  })

  describe("qty in cart, change number input", () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it("calls setQty", () => {
      const story = mountedStories["qty in cart"]
      const { qtyInput } = story.elements
      const { setQty } = mockActions
      const mockEvent = { target: { value: 1 } }
      expect(qtyInput.length).toBe(1)
      expect(setQty).not.toHaveBeenCalled()
      qtyInput.simulate("change", mockEvent)
      expect(setQty).toHaveBeenCalled()
      expect(setQty).toHaveBeenCalledWith(mockEvent.target.value)
    })
  })

  describe("sample, click add to cart button", () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it("calls addToCart", () => {
      const story = mountedStories["sample"]
      const { addToCartButton } = story.elements
      const { addToCart } = mockActions
      expect(addToCartButton.length).toBe(1)
      expect(addToCart).not.toHaveBeenCalled()
      addToCartButton.simulate("click")
      expect(addToCart).toHaveBeenCalled()
    })
  })
})
