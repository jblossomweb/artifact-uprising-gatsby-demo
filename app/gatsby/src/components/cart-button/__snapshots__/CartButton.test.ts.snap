// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/cart-button invalid qty matches snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
  data-testid="addToCartButton"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Add to Cart 
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
      />
    </svg>
  </span>
</button>
`;

exports[`components/cart-button qty in cart matches snapshot 1`] = `
<span
  className="sc-AxjAm igVFIK"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInputBase-adornedStart MuiInputBase-adornedEnd"
    data-testid="qtyInput"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
      data-testid="subtractButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13H5v-2h14v2z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <input
      className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
      max={99}
      min={0}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value={1}
    />
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
      data-testid="addButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
   
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
    />
  </svg>
</span>
`;

exports[`components/cart-button sample matches snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
  data-testid="addToCartButton"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Add to Cart 
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
      />
    </svg>
  </span>
</button>
`;
